src/main/java/gov/cms/madie/terminology/service/FhirTerminologyService.java

```diff
@@ -239,4 +238,102 @@ private void updateOrInsertAllCodeSystems(List<CodeSystem> codeSystemList) {
       }
     }
   }
+
+  public List<Code> retrieveCodesList(List<Map<String, String>> codeList, String apiKey) {
+    return codeList.stream()
+        .map(
+            code -> {
+              List<CodeSystemEntry> codeSystemEntries = mappingService.getCodeSystemEntries();
+              String codeName = code.get("code");
+              String codeSystemName = code.get("codeSystem");
+              String oid = code.get("oid") != null ? code.get("oid").replaceAll("'|'", "") : null;
+
+              Optional<Map.Entry<String, String>> mappedVersion =
+                  oid != null
+                      ? mapToFhirVersion(code.get("version"), oid, codeSystemEntries)
+                      : null;
+
+              if (mappedVersion != null) {
+                String vsacVersion = mappedVersion.get().getKey();
+                String fhirVersion = mappedVersion.get().getValue();
+
+                return retrieveCodes(
+                    codeName, codeSystemName, vsacVersion, fhirVersion, oid, apiKey);
+              }
+
+              return retrieveCodes(codeName, codeSystemName, null, null, oid, apiKey);
+            })
+        .collect(Collectors.toList());
+  }
+
+  private Optional<Map.Entry<String, String>> mapToFhirVersion(
+      String version, String oid, List<CodeSystemEntry> codeSystemEntries) {
+
```

Comment:
```suggestion
    if (oid == null) {
      return Optional.empty();
    }
```
Recommend moving the null check to here to avoid having to check for null before every usage of this method (which, so far, there's only 1).

Reply:
I moved it to here to make it more generic